pennylane/devices/qubit/initialize_state.py

```diff
@@ -0,0 +1,94 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to prepare a state."""
+
+from typing import List
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.interfaces import INTERFACE_MAP
+from pennylane.operation import Operator
+
+INTERFACE_MAP = INTERFACE_MAP.copy()
+del INTERFACE_MAP[None]
+del INTERFACE_MAP["auto"]
+
+
+def _get_padded_wire_order(num_wires, operations):
+    """
+    Gets all distinct wires in ``operations``, and pads that list
+    with new labels if needed.
+
+    Args:
+        num_wires (int): The required number of wires in the returned wire_order
+        operations (List[Operator]): The inputted set of operations to extract wires from
+
+    Returns:
+        wire_order (List[int]): A wire order to use in computations to ensure consistency
+
+    Raises:
+        ValueError: If too many distinct wires are found in the provided list of operations
+    """
+    all_wires = qml.wires.Wires.all_wires([op.wires for op in operations])
+    num_prep_wires = len(all_wires)
+    if num_prep_wires > num_wires:
+        raise ValueError(
+            f"Expected no more than {num_wires} distinct wires across all prep_operations, got {all_wires}"
+        )
+    wire_order = all_wires.tolist()
+    if num_prep_wires < num_wires:  # pad `wire_order` to match `num_wires`
+        delta = num_wires - num_prep_wires
+        if int_wires := [w for w in wire_order if isinstance(w, int)]:
+            first_new_wire = max(int_wires) + 1
+            wire_order.extend(range(first_new_wire, first_new_wire + delta))
+        else:
+            wire_order.extend(range(delta))
+    return wire_order
+
+
+def initialize_state(
+    num_wires: int, prep_operations: List[Operator] = None, framework: str = "autograd"
+) -> np.ndarray:
+    """
+    Initialize a state vector given some preparation operations.
+
+    Args:
+        num_wires (int): The number of wires in the state being initialized
+        prep_operations (List[Operator]): A sequence of operations to apply to prepare an initial state
+        framework (str): The machine learning framework to use when preparing the initial state
+
+    Returns:
+        array[complex]: The initialized state vector
+
+    Raises:
+        QuantumFunctionError: If an unknown machine learning framework is provided
+        ValueError: If too many distinct wires are found in the provided list of prep operations
+    """
+    if framework not in INTERFACE_MAP:
+        raise qml.QuantumFunctionError(
+            f"Unknown framework {framework}. Interface must be one of {list(INTERFACE_MAP)}."
+        )
+    framework = INTERFACE_MAP[framework]
+    if framework == "tf":
+        framework = "tensorflow"
```

Comment:
If the math module doesn't like `tf`, maybe we should either:

* change `INTERFACE_MAP` and usage of `"tf"` in PennyLane source code
* Update the math module to accept `"tf"`.

Both out of scope of this PR, but I can see it being slightly problematic and error-prone if we have two ways of specifying an interface that work in different situations.

Reply:
I was planning on bringing this up in a subsequent PR, but yep. More generally, is `INTERFACE_MAP` doing what it should? should we replace it in this new execution pipeline and be done with the old way? My goal in this PR was to have something functional with minimal changes, and get it optimal in another one

Reply:
This is a result of:

- PennyLane (from v0.3 when TF support was added) using `interface="tf"` as part of the public API to use TensorFlow from the QNode

- Autoray being added in v0.something (maybe ~0.13?) to reduce overhead for developers. And Autoray using `like="tensorflow"`.

We since added `interface="tensorflow"` as an option (I believe) to the QNode.

Reply:
it works in QNode, but it maps "tensorflow" to "tf" internally. Unfortunately, I'm using the resulting _values_ of the map rather than the supported _keys_ (for everything but tf, this makes sense), so for tf I have to un-map it back to "tensorflow" 😄

Reply:
ah 😆